### PR TITLE
added calf ration manager class

### PR DESCRIPTION
**Pull request by JoeWaddell**
Straightforward PR for Milestone 9 in the Ration Refactor Design Doc: https://docs.google.com/document/d/1AvB1Lee54gf_6vsOo03BolMJ4qabZWoqJP9QkDD-yig/edit

Here we simply want to add a class to the methods in calf_ration.py, which is separate enough from other ration methods that it is worth isolating in its own file. Some of the nutrient requirement calculations are not actually yet used, but remain because they may be worth reporting in the future. 

calc_requirements has been split into two methods per Issue #97, tests also added.

### CODE REVIEW

**Comment by JoeWaddell** [comment]
This should close the following issue: https://github.com/RuminantFarmSystems/MASM/issues/97

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/ration/calf_ration.py:88]: Should we make these IDs constants maybe? These values seem to be static so this may be a good opportunity to use constants.

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/life_cycle/calf.py:52]: I'm sure that this logic is correct, but I think this is a rather weak check. Is "body_weight" being in args suffice to know that we can start assigning calf values? What if the "body_weight" is of the wrong type? Would be open to having a discussion about this check.

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/life_cycle/calf.py:49]: I think this should at the very least be made plural, to `self.animal_intakes`, and preferably, if this is a calf-specific value, be named `self.calf_intakes`.

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/life_cycle/calf.py:149]: Refer to comment on line 49 in calf.py regarding `self.animal_intake`

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/life_cycle/calf.py:150]: Refer to comment on line 49 in calf.py regarding `self.animal_intake`

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/ration/calf_ration.py:65]: Refer to comment on line 49 in calf.py regarding `self.animal_intake`

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/ration/calf_ration.py:126]: In general, the more we can avoid nesting if statements, the better. I think the literature requires a decent amount of nesting here, but one fix that can be made is initializing `t_factor` outside of the if statement. This would be done by adding `t_factor = 0` before line 124, and deleting the two innermost else clauses within the if statement (lines 129, 130, 136, 137).

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/ration/calf_ration.py:196]: I think that this should be renamed to `feed_dict` or `feed_amount_dict`, something around those lines.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Fjordrunner) [inline RUFAS/routines/animal/ration/calf_ration.py:289]: As referenced earlier, I think we should choose a different name than "animal_intake" here.

**Review by Fjordrunner — DISMISSED** [review]
Overall, this looks good, but I do think there is some room for improvement. I'd be happy to have discussions about some of the comments because some of them might require some restructuring but overall very good, sorry for being nitpicky!

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/life_cycle/calf.py:52]: This might be something we refactor with Animal Manager, since it's so dependent on how the calf objects are initialized. These methods are mostly the original functions - in this PR I split one big one into two (to handle an old issue), though we're still not truly using the calf ration methods in full. I think changes here are mostly outside the scope of the current PR (just had to make changes for linting, and to use the newly split calf ration methods)

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/ration/calf_ration.py:88]: What it should actually do under normal conditions is to pull Feed IDs from the Feed input JSON, but this is a hardcoded placeholder until final decisions are made on Calf ration formulation, likely with Kristan's insight into what RuFaS needs in the short and long term.

We may also revisit this during the week while we're designing the Feed module refactor.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/ration/calf_ration.py:126]: Good call!

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/ration/calf_ration.py:196]: Also going to hold off on changes until the Feed Redesign/Refactor has made more decisions related to this! Right now it's an instance of Feed that we pull the Calf-specific feeds from, but that structure may change.

**Comment by KFosterReed** [comment]
Simon has some great suggestions for future work but since, like Joe said, we actually aren't using most of the methods in this PR, I think making more substantial changes that influence other parts of the Feed or Animal module is outside the scope of this work. Could we create an issue summarizing Simon's suggestions? 

The changes still work as expected for me with the static /fixed calf 'ration'. My only question before approving is to ask about a plan to update the feed IDs with the new feed library. Is there one? It will just be to change the hard coded feed ids to the new feed library one. Can we include it with the feed library update?

**Comment by JoeWaddell** [comment]
> My only question before approving is to ask about a plan to update the feed IDs with the new feed library.

Great question: once PR #794 is ready and merged, we will need to switch to the new IDs.

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (KFosterReed) [reply RUFAS/routines/animal/life_cycle/calf.py:49]: I think it makes sense to me to leave it not plural. This value is for the intake for one animal on one day. We would refer to this as the calf's intake or her daily intake. 

Question about using the term calf vs. animal - would this attribute be inherited by the other animal classes as the calf graduates to a heifer 1, 2, and so on? If so, seems like it should stay as animal. If not, calf makes sense to me.

**Review by KFosterReed — APPROVED** [review]
I think this is a good step forward for refactoring this part of the model. Suggestions for future improvements can be recorded in issues for when we will actually use the calf diet formulation.

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/life_cycle/calf.py]: Can parameter descriptions be added for `feed` and `temp` here?

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/life_cycle/calf.py:173]: I'm curious, was this change made by the Black formatter?

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py:10]: Should this be included? I had a conversation with Pooya yesterday and he mentioned that we are trying to remove all the shebangs from RuFaS.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]: Units

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
        energy_allow_gain = 0
        if ne_gain >= 0:
            energy_allow_gain = math.exp(0.833 * math.log((1.19 * ne_gain)/(0.69 * calf.body_weight ** 0.355)))
```

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [reply RUFAS/routines/animal/ration/calf_ration.py:126]: You could even go so far as to put this logic in its own method, but that may clutter up the file/PR too much to be worth it.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]: I think it would be good to list the keys returned here and describe what the abbreviations mean. Also add units where appropriate.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
        Returns "optimized" calf ration.

        Notes
        -----
        Use of calf nutrient requirements not implemented, hardcoded ration is returned.

        Returns
        -------
        Dictionary of formulated ration.
```

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
        Calculate dietary intake and nutrient requirements for the calf.

        Parameters
        ----------
        calf : Calf
            Calf object.
        feed : Dict[str, float]
            Instance of Feed class defined in feed.py.
        temp : float
            Average temperature of the simulation day.
        animal_intake : Dict
            Dictionary of calculated intake values.

        Returns
        -------
        Dict[str,  Dict[str, Any]]
            Dictionary of requirement values and methods.
```

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
        Hardcoded calf ration.
```

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
            Dictionary of formulated ration.
```

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
        Calculating calf intake values

        Parameters
        ----------
        calf : Calf
            Calf object.
        feed : Dict[str, float]
            Instance of Feed class defined in feed.py.
        wean_day : int
            Wean day of the calf.
        wean_length : int
            Wean length of the calf.
        milk_type : string
            Either "whole" or "replacer".

        Returns
        -------
        Dict[str, float]
            Dictionary of intake values.

        """
```

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]: Same comment as above for adding dictionary keys and units.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]: I think these descriptions could be made more specific in a couple of ways: 
1. Change the descriptions so that they include what weaning is, like "Wean day of calf" => "Day calf is separated from its mother". For anyone with the most basic understanding of animal science, I'm sure it would be unnecessary to redefine weaning for them. But for people with zero knowledge of this stuff, it helps to have these terms more spelled out.
2. `wean_day` doesn't need units specified, but it is not totally clear how days are being used here. Is it days after the calf's birth, or julian day of the year on which the calf is weaned, or something else? Ex. if it is the first option, then something like "Wean day of calf" => "Number of days after birth that calf is weaned" would make it clearer.
3. `wean_length` should have units - pretty sure it is days but I think should still be specified anyways.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py:125]: Finding a way to eliminate these inline comments would be good - there are a lot of them.

For the equation references, the docstring could have:
```
    References
    ----------
    See Animal pseudocode equations from [A.1B.C.1] to [A.1B.H.5].
```
If removing those really specific mappings would make things too ambiguous though, an alterative would be to separate each equation into its own method and reference the specific equation in the docstring there. That solution would really blow this file and the test file though.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/calf_ration.py]:
```suggestion
    @classmethod
```

File: tests/test_animal.py
Comment (Isgotkowitz) [inline tests/test_animal.py]:
```suggestion
def test_get_ration():
```

File: tests/test_animal.py
Comment (Isgotkowitz) [inline tests/test_animal.py:1]: Tests in here look good, but all tests for `CalfRationManager` should be moved to into a dedicated file (`test_calf_ration.py`), which should have the path `tests/test_animal_module/ration/test_calf_ration.py`. This PR seems like a good opportunity for make this change, but you may have a timeline in mind for how this file should be broken down.

File: tests/test_animal.py
Comment (Isgotkowitz) [inline tests/test_animal.py]: Right now the docstrings for these tests specify what they are testing but not what they are testing for, which I think they should.
```suggestion
    """Tests the optimize function to verify it returns a correctly optimized calf ration."""
```

File: RUFAS/scenario_manager.py
Comment (Isgotkowitz) [inline RUFAS/scenario_manager.py]: Don't forget to change this back before merging.

File: README.md
Comment (Isgotkowitz) [inline README.md:3]: Nice

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/life_cycle/calf.py]:
```suggestion
```

**Review by Isgotkowitz — APPROVED** [review]
Overall looks great Joe.

Comments mostly revolve around docstrings/comments.

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/life_cycle/calf.py:173]: It was: this and a few others wound up with slightly wacky-looking formatting.

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/ration/calf_ration.py:10]: Great question, I'll ask

File: RUFAS/routines/animal/ration/calf_ration.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/ration/calf_ration.py:125]: This is a great recommendation: it's something we'll have to do across all nutrient requirement calculation methods (here and in other files) for consistency, but it will certainly clean up the code nicely.

On revision, I'm going to do this in a separate PR as a general "comment cleanup" to also ensure that all Pseudocode is up to date.

File: tests/test_animal.py
Comment (JoeWaddell) [reply tests/test_animal.py:1]: I'll also push this to the final PRs in this ration refactor, since those do a lot of test additions/rearranging.

**Pull request merged**
